nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java

```diff
@@ -0,0 +1,216 @@
+
+        FlowFile outgoingFlowFile;
+
+        String databaseName = context.getProperty(DB_NAME).getValue();
+        String adxQuery = null;
+        KustoResultSetTable kustoResultSetTable;
+
+        if ( context.hasIncomingConnection() ) {
+            FlowFile incomingFlowFile = session.get();
+
+            if ( incomingFlowFile == null && context.hasNonLoopConnection() ) {
+                String message = "Incoming connection exists but Incoming flow file is null";
+                getLogger().error(message);
+                incomingFlowFile = session.putAttribute(incomingFlowFile, ADX_QUERY_ERROR_MESSAGE, message);
+                session.transfer(incomingFlowFile, RL_FAILED);
+                return;
+            }
+
+            if ( incomingFlowFile != null && incomingFlowFile.getSize() == 0 ) {
+                if ( context.getProperty(ADX_QUERY).isSet() ) {
+                    adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = session.putAttribute(incomingFlowFile, ADX_QUERY_ERROR_MESSAGE, message);
+                    session.transfer(incomingFlowFile, RL_FAILED);
+                    return;
+                }
+            } else {
+                try {
+                    adxQuery = getQuery(session, StandardCharsets.UTF_8, incomingFlowFile);
+                } catch(IOException ioe) {
+                    getLogger().error("Exception while reading from FlowFile " + ioe.getLocalizedMessage(), ioe);
+                    throw new ProcessException(ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+
+        } else {
+            outgoingFlowFile = session.create();
+            adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        try {
+            //execute Query
+            kustoResultSetTable = executeQuery(databaseName,adxQuery);
+            List<List<Object>> tableData = kustoResultSetTable.getData();
+
+            String json = tableData.size() == 1 ? objectMapper.writeValueAsString(tableData.get(0)) : objectMapper.writeValueAsString(tableData);
+
+            if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Query result {} ", new Object[] {tableData});
+            }
+
+            ByteArrayInputStream bais = new ByteArrayInputStream(json.getBytes());
+            session.importFrom(bais, outgoingFlowFile);
+            bais.close();
+
+            //if no error then
+            outgoingFlowFile = session.putAttribute(outgoingFlowFile, ADX_EXECUTED_QUERY, String.valueOf(adxQuery));
+            session.transfer(outgoingFlowFile, RL_SUCCEEDED);
+        }catch (OutOfMemoryError ex){
```

Comment:
Yes I do understand that it is not recommended to catch OOM. But when I tried to check in local it throws an error instead of exception. Had it been exception, the message body could have been used to understand if the query limits are hit. Hence thought of catching OOM. Not sure if it is specific to my local env. The other possibility I see is catching Throwable

```
} catch (DataServiceException | DataClientException | IOException e) {
            getLogger().error("Exception occurred while reading data from ADX ", e);
            session.transfer(outgoingFlowFile, RL_FAILED);
        } catch (Throwable ex){
            getLogger().error("Exception occurred while reading data from ADX : Query Limits exceeded : Please modify your query to fetch results below the kusto query limits ", ex);
            session.transfer(outgoingFlowFile, RL_FAILED);
        }
```

Reply:
```
          val isRowLimitHit = ExceptionUtils.getRootCauseStackTrace(exception).contains("Query execution has exceeded the allowed limits")
            if (isUserOptionForceSingleMode && !isRowLimitHit) {
              // Expected behavior for Issue#261
              throw exception
            } else {

```
something like this

Reply:
identified the source of OOM. issue was with .getPrimaryResults method. changed the implementation to catch the KustoQueryLimits exceeded exception.